pubsub/google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py

```diff
@@ -0,0 +1,382 @@
+
+        Returns:
+            float: The load value.
+        """
+        if self._leaser is None:
+            return 0
+
+        return max([
+            self._leaser.message_count / self._flow_control.max_messages,
+            self._leaser.bytes / self._flow_control.max_bytes,
+        ])
+
+    def add_close_callback(self, callback):
+        """Schedules a callable when the manager closes.
+
+        Args:
+            callback (Callable): The method to call.
+        """
+        self._close_callbacks.append(callback)
+
+    def maybe_pause_consumer(self):
+        """Check the current load and pause the consumer if needed."""
+        if self.load >= 1.0 and not self._consumer.is_paused:
+            _LOGGER.debug(
+                'Message backlog over load at %.2f, pausing.', self.load)
+            self._consumer.pause()
+
+    def maybe_resume_consumer(self):
+        """Check the current load and resume the consumer if needed."""
+        # If we have been paused by flow control, check and see if we are
+        # back within our limits.
+        #
+        # In order to not thrash too much, require us to have passed below
+        # the resume threshold (80% by default) of each flow control setting
+        # before restarting.
+        if not self._consumer.is_paused:
+            return
+
+        if self.load < self.flow_control.resume_threshold:
+            self._consumer.resume()
+        else:
+            _LOGGER.debug('Did not resume, current load is %s', self.load)
+
+    def send(self, request):
+        """Queue a request to be sent to the RPC."""
+        self._rpc.send(request)
+
+    def open(self, callback):
+        """Begin consuming messages.
+
+        Args:
+            callback (Callable[None, google.cloud.pubsub_v1.message.Messages]):
+                A callback that will be called for each message received on the
+                stream.
+        """
+        if self.is_active:
+            raise ValueError('This manager is already open.')
+
+        if self._closed:
+            raise ValueError(
+                'This manager has been closed and can not be re-used.')
+
+        self._callback = functools.partial(_wrap_callback_errors, callback)
+
+        # Start the thread to pass the requests.
+        self._dispatcher = dispatcher.Dispatcher(self, self._scheduler.queue)
+        self._dispatcher.start()
+
+        # Start consuming messages.
+        self._rpc = bidi.ResumableBidiRpc(
+            start_rpc=self._client.api.streaming_pull,
+            initial_request=self._get_initial_request,
+            should_recover=self._should_recover)
+        self._rpc.add_done_callback(self._on_rpc_done)
+        self._consumer = bidi.BackgroundConsumer(
+            self._rpc, self._on_response)
+        self._consumer.start()
+
+        # Start the lease maintainer thread.
+        self._leaser = leaser.Leaser(self)
+        self._leaser.start()
+
+    def close(self, reason=None):
```

Comment:
It can be called from multiple threads (namely, it could be closed by the calling `future.cancel` and `on_rpc_done` at the same time). Thankfully, this whole method is guarded by a lock (`with self._closing:`) which forces all calls to this to be handled serially regardless of concurrency. So the "first" caller will actually close and the second will hit the early out branch.

Reply:
I totally missed the lock :|